fuel-vm/src/interpreter/executors/main.rs

```diff
@@ -43,33 +52,53 @@ impl<T> Interpreter<PredicateStorage, T> {
     ///
     /// This is not a valid entrypoint for debug calls. It will only return a `bool`, and not the
     /// VM state required to trace the execution steps.
-    pub fn check_predicates<Tx>(checked: Checked<Tx>, params: ConsensusParameters, gas_costs: GasCosts) -> bool
+    pub fn check_predicates<Tx>(
+        checked: Checked<Tx>,
+        params: ConsensusParameters,
+        gas_costs: GasCosts,
+    ) -> Result<PredicatesChecked, PredicateVerificationFailed>
     where
-        Tx: ExecutableTransaction,
+        Tx: ExecutableTransaction + fuel_tx::field::GasLimit,
         <Tx as IntoChecked>::Metadata: CheckedMetadata,
     {
-        let mut vm = Interpreter::with_storage(PredicateStorage::default(), params, gas_costs);
-
         if !checked.transaction().check_predicate_owners() {
-            return false;
+            return Err(PredicateVerificationFailed);
         }
 
-        #[allow(clippy::needless_collect)]
+        let mut vm = Interpreter::with_storage(PredicateStorage::default(), params, gas_costs);
+
         // Needed for now because checked is only freed once the value is collected into a Vec
-        let predicates: Vec<RuntimePredicate> = checked
-            .transaction()
-            .inputs()
-            .iter()
-            .enumerate()
-            .filter_map(|(idx, _)| RuntimePredicate::from_tx(&params, checked.transaction(), idx))
+        #[allow(clippy::needless_collect)]
+        let predicates: Vec<_> = (0..checked.transaction().inputs().len())
+            .filter_map(|i| RuntimePredicate::from_tx(&params, checked.transaction(), i))
             .collect();
 
-        predicates
-            .into_iter()
-            .fold(vm.init_predicate(checked), |result, predicate| -> bool {
-                // VM is cloned because the state should be reset for every predicate verification
-                result && vm.clone()._check_predicate(predicate)
-            })
+        // Since we reuse the vm objects otherwise, we need to keep the actual gas here
+        let tx_gas_limit = *checked.transaction().gas_limit();
+        let mut remaining_gas = tx_gas_limit;
+
+        vm.init_predicate(checked);
+
+        for predicate in predicates {
+            // VM is cloned because the state should be reset for every predicate verification
+            let mut vm = vm.clone();
+
+            vm.context = Context::Predicate { program: predicate };
+            vm.registers[REG_GGAS] = remaining_gas;
+            vm.registers[REG_CGAS] = remaining_gas;
+
+            if !matches!(vm.verify_predicate(), Ok(ProgramState::Return(0x01))) {
+                return Err(PredicateVerificationFailed);
+            }
+
+            remaining_gas = vm.registers[REG_GGAS];
+        }
+
+        Ok(PredicatesChecked {
+            gas_used: tx_gas_limit
+                .checked_sub(remaining_gas)
+                .expect("Bug! Execution increased available gas"),
```

Comment:
We should use the `Bug` error type here instead of expect: https://github.com/FuelLabs/fuel-vm/blob/master/fuel-vm/src/error.rs#L292